modules/biobank/php/specimenendpoint.class.inc

```diff
@@ -0,0 +1,123 @@
+<?php declare(strict_types=1);
+/**
+ * This implement the Specimen endpoint class for the Biobank Module.
+ *
+ * PHP Version 7.2
+ *
+ * @category   API
+ * @package    Loris
+ * @subpackage Biobank
+ * @author     Henri Rabalais <henri.rabalais@mcin.ca>
+ * @license    http://www.gnu.org/licenses/gpl-3.0.text GPLv3
+ * @link       http://www.github.com/aces/Loris/
+ */
+namespace LORIS\biobank;
+use \Psr\Http\Message\ServerRequestInterface;
+use \Psr\Http\Server\RequestHandlerInterface;
+use \Psr\Http\Message\ResponseInterface;
+
+/**
+ * A class for handling the /biobank/specimens/ endpoint.
+ *
+ * PHP Version 7.2
+ *
+ * @category   Biobank
+ * @package    Main
+ * @subpackage Biobank
+ * @author     Henri Rabalais <henri.rabalais@mcin.ca>
+ * @license    http://www.gnu.org/licenses/gpl-3.0.text GPLv3
+ * @link       http://www.github.com/aces/Lors/
+ */
+class SpecimenEndpoint implements RequestHandlerInterface
+{
+    /**
+     * Returns true if user has acces to this Controller.
+     *
+     * @param \User $user The user whose access is being checked
+     *
+     * @return bool
+     */
+    function _hasAccess(\User $user) : bool
+    {
+        return true;
+    }
+
+    /**
+     * Return which methods are supported by this endpoint.
+     *
+     * @return array supported HTTP methods
+     */
+    protected function allowedMethods() : array
+    {
+        return ['GET', 'PUT', 'POST'];
+    }
+
+    /**
+     * This function passes the request to the handler. This is necessary since
+     * the Endpoint bypass the Module class.
+     *
+     * XXX: This function should be extracted to a parent class. 
+     * 
+     * @param ServerRequestInterface  $request The PSR7 request.
+     * @param RequestHandlerInterface $handler The request handler.
+     *
+     * @return ResponseInterface The outgoing PSR7 response.
+     */
+    public function process(
+        ServerRequestInterface $request,
+        RequestHandlerInterface $handler
+    ) : ResponseInterface {
+        return $handler->handle($request);
+    }
+
+    /**
+     * Handles Controller requests
+     *
+     * @param ServerRequestInterface $request The incoming PSR7 request
+     *
+     * @return ResponseInterface The outgoing PSR7 response
+     */
+    public function handle(ServerRequestInterface $request) : ResponseInterface
+    {
+        $response;
+        $db       = \Database::singleton();
+        $user     = $request->getAttribute('user');
+        $specCont = new SpecimenController($db, $user);
+
+        try {
+            // XXX: Not sure if I'm using this file uploader properly. John S.
+            // or Xavier please check this. Perhaps this should be sent to an
+            // entirely different endpoint and not handled within the specimen
```

Comment:
I think it makes sense to me that a specimen upload would be a POST to a specimen endpoint. I'm not super knowledgeable about best practices here though.

Reply:
Okay, thanks!